packages/nimble-components/src/table/specs/cell-state-when-scrolling-hld.md

```diff
@@ -0,0 +1,119 @@
+-   Pending input (for editable controls in table cells)
+-   Open popups (action menu or select/combobox cells whose popup is open)
+
+## Links To Relevant Work Items and Reference Material
+
+[#997: Table virtualization: Clear/Commit (and potentially save/restore) row/cell state when scrolling](https://github.com/ni/nimble/issues/997)
+
+## Implementation / Design
+
+When the user scrolls the table, we will:
+
+-   If a table cell contains an active/focused control, we will notify the column plugin that the cell is being blurred/recycled via a new API. The general expectation is that editable controls will commit any pending updates, then blur the focused the control.
+-   If a cell action menu is open (in which case a menu item is focused), we'll close the associated menu via the menu button.
+
+We will not:
+
+-   Detect and deselect selected table text on scroll. This means that if table text is selected and the user scrolls, text will still be selected (potentially different text). We don't consider that to be a primary table interaction, and it shouldn't harm functionality (it's easy to resolve for the user by clicking off to deselect text). It's also the same behavior as the current `sl-grid`/`smart-table` used in SystemLink Enterprise.
+    -   See the "Alternative Implementations" section for additional information on this decision.
+-   Try and re-apply any state to the re-bound rows/cells after the scroll. That means that we won't re-focus the previously focused control in a cell / re-open an action menu, after a scroll operation.
+
+**Prototype:**  
+See [the prototype branch](https://github.com/ni/nimble/compare/main...table-cell-state-custom-elements-1) and the [prototype table Storybook](https://60e89457a987cf003efc0a5b-azrrwyldta.chromatic.com/?path=/story/table--table&args=data:LargeDataSet) to illustrate the concepts discussed in the following sections.  
+As we don't yet support editable column types, the prototype Storybook updates the Last Name column to show text which can be focused by clicking on it. After clicking a Last Name cell value, you'll see the text get a green border as its focused styling. Once you scroll the table vertically, the green border goes away, and you'll see a `console.log` indicating that the new cell blur API was used.
+
+### Blur Focused Controls in Cells
+
+In this case, `document.activeElement` will be the Nimble `Table`, and `table.shadowRoot.activeElement` will be a Nimble `TableRow`. We can recursively look at the active element's `shadowRoot.activeElement`, starting from the TableRow, and stop when we reach a `BaseCellElement` (the custom element in table cells implemented by column plugin authors) or `null` (see [prototype](https://github.com/ni/nimble/compare/main...table-cell-state-custom-elements-1#diff-e2bfd4eda0a3c89f54b4e09624e6cfd5a68ea2f14c0e79f717eacce38ec5982bR124)).
+
+If we found a focused `BaseCellElement`, then we will call the new API `onBeforeBlur()` on it.  
```

Comment:
I'm struggling to come up with a good alternative, but I'm not a huge fan of `onBeforeBlur()` for this function. At least to me, `onBeforeBlur()` means that the element is about to be blurred so I should take any actions I need to before that blur occurs. However, what is really happening is that I am being told to blur the element myself.

What about something like `blurCell`?

Reply:
I think I see the direction you were going. `blurCell` is like an action, you are telling it to blur. Like the focus method, or the actual blur method lol.

In that vein one option is we actually tell clients that to respond to blur behaviors you need to override the blur method. I think that kinda maybe even makes sense. Like calling the blur method on an input element with an uncommited value commits the value and then removes keyboard focus. Can try it on the mdn page adding a change listener in devtools: https://developer.mozilla.org/en-US/docs/Web/API/HTMLElement/blur

The other web component lifecycle behaviors are named `*Callback` (connectedCallback, disconnectedCallback, attributeChangedCallback). Maybe a `*TableCallback` to emphasize the lifecycle nature of it. And I like the direction of "what is the table doing" which aligns more with the lifecycle nature, i.e. connected/disconnected/attributeChanged are what is the DOM API doing while `*TableCallback` is what is the table doing. Maybe `beforeRecycleTableCallback`, `beforeBlurTableCallback`?

I think I kind of like the hey we will call `blur()` on you when we want to remove focus so commit your thing at that point approach. I think the behavior that blur on native text fields commited the value and caused it to lose focus, which sounds like exactly what we want, is making me lean to that direction. But if we want more control, i.e. the need to pass different parameters then I'd lean to the `*TableCallback` approach.

Reply:
Another thought that maybe overriding blur (and calling super) answers. What should have focus after the blur?
The contract of the blur method is that it is the viewport that has focus:
![image](https://user-images.githubusercontent.com/1588923/224858613-9c5a1150-33d8-463c-9224-a7b8e2998568.png)
When you page down in a keyboard I'd expect a new cell cell to be focused, not for the table to lose focus. So maybe overriding blur is the wrong behavior and the name blur is a bit misleading / unaligned with the spec definition of blur.

And I think I remember that this is more precise than recycling. This is a callback specific to a cell that is about to be recycled that has focus losing focus and then being recycled. Many other cells will be recycled at the same time but this is specifcally a focused cell about to be recycled.

Option 3: `beforeFocusChangeTableCallback`. It's not a "blur" because focus isn't going to the viewport, it's changing to something else in the table (I'm assuming based on the above discussion once keyboard interactions are implemented). It's different from other lifecycle `*Callback` because it happens before not after.

Reply:
This is actually getting really close to the naming from one of your original proposals of `onBeforeFocusedCellRecycled` lol

Option 4: beforeFocusedCellRecycleCallback, beforeFocusedCellRecycle, beforeFocusedRecycle, beforeFocusedRecycleCallback, focusedRecycleCallback, focusedRecycleTableCallback

I think I like focusedRecycleCallback the most and leaves room for a recycleCallback if needed in the future...

Reply:
Updated to `focusedRecycleCallback` naming.  
I agree with keeping `blur` out of the name since we won't be the ones calling `blur` ourselves, the column plugins will.